src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs

```diff
@@ -274,19 +312,24 @@ internal static FilterExpression Parse(string filterString, out FastFilter? fast
     internal bool Evaluate(Func<string, object?> propertyValueProvider)
     {
         ValidateArg.NotNull(propertyValueProvider, nameof(propertyValueProvider));
-        bool filterResult = false;
-        if (null != _condition)
-        {
-            filterResult = _condition.Evaluate(propertyValueProvider);
-        }
-        else
-        {
-            // & or | operator
-            bool leftResult = _left!.Evaluate(propertyValueProvider);
-            bool rightResult = _right!.Evaluate(propertyValueProvider);
-            filterResult = _areJoinedByAnd ? leftResult && rightResult : leftResult || rightResult;
-        }
-        return filterResult;
+
+        return IterateFilterExpression<bool>((current, result) =>
+         {
+             bool filterResult = false;
+             if (null != current._condition)
+             {
+                 filterResult = current._condition.Evaluate(propertyValueProvider);
+             }
+             else
+             {
+                 // & or | operator
+                 bool rightResult = current._right != null ? result.Pop() : false;
+                 bool leftResult = current._left != null ? result.Pop() : false;
+                 filterResult = current._areJoinedByAnd ? leftResult && rightResult : leftResult || rightResult;
+             }
+             return filterResult;
+         });
+
```

Comment:
Nit: remove blank line + fix indentation for the curly braces.

```suggestion
        return IterateFilterExpression<bool>((current, result) =>
        {
             bool filterResult = false;
             if (null != current._condition)
             {
                 filterResult = current._condition.Evaluate(propertyValueProvider);
             }
             else
             {
                 // & or | operator
                 bool rightResult = current._right != null ? result.Pop() : false;
                 bool leftResult = current._left != null ? result.Pop() : false;
                 filterResult = current._areJoinedByAnd ? leftResult && rightResult : leftResult || rightResult;
             }
             return filterResult;
        });
```